### PR TITLE
Improved Animations and bugfix for #4

### PR DESCRIPTION
**Pull request by Simbaclaws**
Dear @kolkov ,

I've succesfully improved the animations of the plugin by making sure that whatever thumbnail is selected the animation would still be played for the corresponding selection. Including the fix for https://github.com/kolkov/ngx-gallery/issues/4.

However a small new bug arises when using this code. I haven't yet found a solution for it but I doubt it's a big issue. Apparently when using the slide animation there is a small white space between the slides when going to a different slide. This wasn't there before and is created by my pull request. I'm guessing it's got something to do with this code:

```
  addActive(index, image) {
    setTimeout(()=>{
      if (this.selectedIndex == index) {
        image.classList.add('ngx-gallery-active');
      }
    }, 0);
  }
```

I had to add a timeout in order to trigger the transition animation for the slide effect for the incomming picture. However since there is a 0 ms delay it causes some white space to appear and I haven't yet figured out how to get rid of it.

I'm looking to fix that bug as well, but for now I think this is already a lot better then what the plugin previously could do.

Please take a look to see whether it is good enough for now.

### CODE REVIEW

**Comment by kolkov** [comment]
Thanks for the PR) I try to review this soon as possible.

**Comment by kolkov** [comment]
One more issue. Video continues playing on selecting next frame.

**Comment by Simbaclaws** [comment]
Oh, that is an issue that probably also exists in the current master branch then, oh oh.
Tomorrow I'll try to fix both issues.

**Comment by Simbaclaws** [comment]
Just made sure the video stops playing when switching to a different image or video, it is also reset to the beginning of the video.

I haven't yet found a solution for the white space problem, I find this one hard to solve.

**Comment by kolkov** [comment]
I showed in a comment how you can solve the problem by moving the logic of adding classes to the class method.

**Comment by kolkov** [comment]
This under code review...
image

**Comment by kolkov** [comment]
And you don't need everyday to change path to library if you run build and watch lib and then run start to build application... All refreshes automatically if you change library source code.

**Comment by kolkov** [comment]
And try to use Goland + AngularJS plugin or Webstorm EAP to developing. It is very convenient and intelligent IDEs.

https://blog.jetbrains.com/go/2020/02/06/welcome-to-the-goland-2020-1-eap/

https://confluence.jetbrains.com/display/WI/WebStorm+EAP

**Comment by Simbaclaws** [comment]
odd, I can't see the code review messages, are they still pending to be added to the code review? I literally don't have them.

I'll try what you said

**Comment by kolkov** [comment]
Thanks!

File: projects/gallery/src/lib/ngx-gallery-image/ngx-gallery-image.component.html [hunk 1]
```diff
@@ -2,11 +2,14 @@
...
-        [ngClass]="{ 'ngx-gallery-active': selectedIndex == image.index,
+    <div #curimage *ngIf="image.type === 'image'" class="ngx-gallery-image"
+        [ngClass]="{ 'ngx-gallery-active-from-left': selectedIndex > currentIndex && selectedIndex == image.index,
+          'ngx-gallery-active-from-right': selectedIndex < currentIndex && selectedIndex == image.index,
           'ngx-gallery-inactive-left': selectedIndex > image.index,
           'ngx-gallery-inactive-right': selectedIndex < image.index,
-            'ngx-gallery-clickable': clickable }"
+          'ngx-gallery-inactive': selectedIndex != image.index && currentIndex != image.index,
+            'ngx-gallery-clickable': clickable,
+            'ngx-gallery-active': addActive(image.index, curimage) }"
```
Comment (kolkov) [inline projects/gallery/src/lib/ngx-gallery-image/ngx-gallery-image.component.html:12]: Maybe we can move the whole class logic into one function?
like this:
```ts
/*
  addActive(index, image) {
    setTimeout(()=>{
      if (this.selectedIndex === index) {
        image.classList.add('ngx-gallery-active');
      }
    }, 0);
  }*/

  getClasses(image, current) {
    let isActive = false;
      if (this.selectedIndex === image.index) {
        isActive  = true;
      }
    return {'ngx-gallery-active-from-left': this.selectedIndex > this.currentIndex && this.selectedIndex === image.index,
      'ngx-gallery-active-from-right': this.selectedIndex < this.currentIndex && this.selectedIndex === image.index,
      'ngx-gallery-inactive-left': this.selectedIndex > image.index,
      'ngx-gallery-inactive-right': this.selectedIndex < image.index,
      'ngx-gallery-inactive': this.selectedIndex !== image.index && this.currentIndex !== image.index,
      'ngx-gallery-clickable': this.clickable,
      'ngx-gallery-active': isActive }
  }
```

**Comment by kolkov** [comment]
At the moment, this commit does not solve the problem with delayed frame animation.

**Comment by kolkov** [comment]
Please check the master branch for changes.

**Comment by Simbaclaws** [comment]
This indeed does not solve the delayed frame animation problem, what it does solve is that images won't dissapear when clicking on the next thumbnail or next arrow repeatedly. A bug I found when trying to find for issues.

I haven't found a way to fix the delayed animation problem just yet but I have fixed a other bug that I saw was appearing.

I will check for the conflicts and fix them.

What do you mean with approach c?

**Comment by kolkov** [comment]
> РЕДАКТИРОВАТЬ: Ах, я вижу, вы изменили мастер ветку, я посмотрю и посмотрю, является ли это лучший подход.
> 
> EDIT2: я проверил основную ветку, ваш подход кажется намного лучше, чем мой. Я думаю, это решено тогда? Должен ли я закрыть этот запрос?
> 
> EDIT4:
> 
> Возможно, неплохо было бы добавить мои последние изменения с помощью логического выражения isAnimating.
> [32fa38b # diff-391906b00a686cfb5efc80c7b07aefbaR181-R201](https://github.com/kolkov/ngx-gallery/commit/32fa38bf0c334dcf2fe338405d74119ba1273bec#diff-391906b00a686cfb5efc80c7b07aefbaR181-R201)
> 
> Это сделало бы невозможным быстрое переключение между миниатюрами, в то время как анимация все еще делает свое дело, не вызывая также появления белой задержки.
> 
> Когда вы внесете эти изменения в репозиторий npm?

Please check all possible problems that we can get by  radically  changing the approach. One head is good, and two is better, as they say. The only thing I see is that we are depriving people of some kind of customization through the use of redefinition in and other possible animations.
What are your thoughts on this approach?
https://blog.thoughtram.io/angular/2017/07/26/a-web-animations-deep-dive-with-angular.html

**Comment by kolkov** [comment]
https://ngx-gallery.kolkov.ru/ live preview with latest changes

**Comment by Simbaclaws** [comment]
is https://ngx-gallery.kolkov.ru live? Because I can see a difference between the master branch and this live preview. The live preview looks way worse then the master branch when doing git clone.

So what you are saying is that the approach you took for animating the next and previous image is something we'd need to do for all different sort of animations?

This is the reason why I wanted to do it with css classes, if you look at my repository you can see that all different types of animations work. Perhaps we might need to do something similar but with angular animations instead. I'll have a look at what I can do to the master branch.

For the problem when clicking multiple times I say maybe use the isAnimating boolean perhaps? Or would this be limiting to the user?

**Comment by kolkov** [comment]
I fix all known issues, and update live example.

**Comment by Simbaclaws** [comment]
Thank you, you're awesome.

I see that the master branch doesn't have a fix for the other animations.
I made a fix for that.

**Comment by Simbaclaws** [comment]
Odd, the conflicts don't seem to resolve....

**Comment by kolkov** [comment]
I think we need to add transition duration to config. And we need to extract transitions to external file in future.

**Comment by kolkov** [comment]
I think we need to close this PR and start another PR with latest changes.

**Comment by Simbaclaws** [comment]
Agreed

**Pull request closed without merging**